File: Linked List/circular/addition/list.py
# Circular Singly Linked List - Addition (Python)
# Insert at head and tail.

class Node:
    def __init__(self, value: int) -> None:
        self.data = value
        self.next: "Node | None" = None


def insert_at_head(head: "Node | None", value: int) -> Node:
    node = Node(value)
    if head is None:
        node.next = node
        return node
    tail = head
    while tail.next is not head:
        tail = tail.next
    node.next = head
    tail.next = node
    return node


def insert_at_tail(head: "Node | None", value: int) -> Node:
    node = Node(value)
    if head is None:
        node.next = node
        return node
    tail = head
    while tail.next is not head:
        tail = tail.next
    tail.next = node
    node.next = head
    return head


def traverse(head: "Node | None") -> None:
    if head is None:
        print()
        return
    cur = head
    while True:
        print(cur.data, end=" ")
        cur = cur.next
        if cur is head:
            break
    print()


if __name__ == "__main__":
    head = None
    head = insert_at_tail(head, 20)
    head = insert_at_tail(head, 30)
    head = insert_at_head(head, 10)
    head = insert_at_tail(head, 40)
    traverse(head)

File: Linked List/circular/addition/list.java
// Circular Singly Linked List - Addition (Java)
// Insert at head and tail.

class Node {
    int data;
    Node next;
    Node(int value) { this.data = value; this.next = null; }
}

public class list {
    static Node insertAtHead(Node head, int value) {
        Node node = new Node(value);
        if (head == null) { node.next = node; return node; }
        Node tail = head;
        while (tail.next != head) tail = tail.next;
        node.next = head;
        tail.next = node;
        return node;
    }

    static Node insertAtTail(Node head, int value) {
        Node node = new Node(value);
        if (head == null) { node.next = node; return node; }
        Node tail = head;
        while (tail.next != head) tail = tail.next;
        tail.next = node; node.next = head;
        return head;
    }

    static void traverse(Node head) {
        if (head == null) { System.out.println(); return; }
        Node cur = head;
        do { System.out.print(cur.data + " "); cur = cur.next; } while (cur != head);
        System.out.println();
    }

    public static void main(String[] args) {
        Node head = null;
        head = insertAtTail(head, 20);
        head = insertAtTail(head, 30);
        head = insertAtHead(head, 10);
        head = insertAtTail(head, 40);
        traverse(head);
    }
}


